docs/design_notes/arguments.md

```diff
@@ -0,0 +1,91 @@
+# Dealing with Module Arguments
+
+## Introduction
+
+In Linen we can define `Module` arguments either as dataclass attributes or as arguments to methods (usually `__call__`).
+Typically the distinction is clear:
+* Completely fixed properties, such as the choice of kernel initializer or number of output features, are hyperparameters and should be defined as dataclass attributes. Typically two Module instances with different hyperparamaters cannot share in a meaningfull way.
+* Dynamic properties, such as input data and top-level "mode switches" like `train=True/False` that should be passed as arguments to `__call__` or another method.
+
+Some cases are however ambigious. Take for example the `Dropout` module.
+We have a number of clear hyperparameters:
+1. The dropout rate
+2. The axes for which a dropout mask is generated
+
+And some clear call time arguments:
+1. The input that should be masked using dropout
+2. The (optional) rng used to sample the random mask
+
+There is however one property that is ambigious -- the `deterministic` property in a Dropout module.
+
+If `deterministic` is `True` no dropout mask is sampled. This is typically used during model evaluation.
+However, if we pass `eval=True` or `train=False` to a top-level Module. The `deterministic` argument needs
+to be applied everywhere and the boolean argument needs to be passed down to all the layers that might use `Dropout`.
+If instead `deterministic` is a dataclass attribute, we might do the following:
+
+```python
+from functools import partial
+from flax import linen as nn
+
+class ResidualModel(nn.Module):
+  drop_rate: float
+
+  @nn.compact
+  def __call__(self, x, *, train):
+    dropout = partial(nn.Dropout, rate=self.drop_rate, deterministic=not train)
+    for i in range(10):
+      x += ResidualBlock(dropout=dropout, ...)(x)
+```
+
+It makes sense to pass `determinstic` to the constructor here because this way we can pass the dropout template to the sub-modules.
+Now the sub-module no longer needs to take care of train vs eval mode and can simply use the `dropout` argument.
+Note that because the dropout layer can only be constructed in the sub-module we can only partially `deterministic` to the constructor but not to `__call__`.
```

Comment:
```suggestion
Note that because the dropout layer can only be constructed in the sub-module we can only partially apply `deterministic` to the constructor but not to `__call__`.
```